src/sctools/barcode.py

```diff
@@ -0,0 +1,134 @@
+
+    def __iter__(self):
+        return iter(self._data)
+
+    def __len__(self):
+        return len(self._data)
+
+    def __getitem__(self, item):
+        return self._data[item]
+
+    def summarize_hamming_distances(self):
+        """returns descriptive statistics on hamming distances between pairs of barcodes"""
+        distances = []
+        for a, b in itertools.combinations(self, 2):
+            distances.append(bin(a ^ b).count('1'))
+        distances.sort()
+
+        return {
+            'minimum': distances[0],
+            '25th percentile': distances[int(len(distances) * .25)],
+            'median': distances[int(len(distances) * .5)],
+            'average': sum(distances) / len(distances),
+            '75th percentile': distances[int(len(distances) * .75)],
+            'maximum': distances[-1]
+        }
+
+    def base_frequency(self, weighted=False):
+        """return the frequency of each base at each position in the barcode set
+
+        :param bool weighted: (Default False) if True, each barcode is counted once for each time it
+          was observed
+        :return np.array: barcode_length x 4 2d array
+        """
+        base_counts_by_position = np.zeros((self._barcode_length, 4), dtype=np.uint64)
+        keys = np.fromiter(self._data.keys(), dtype=np.uint64)
+        values = np.fromiter(self._data.values(), dtype=np.uint64)
+
+        for i in reversed(range(self._barcode_length)):
+            binary_base_representations, counts = np.unique(keys & 3, return_counts=True)
+            if weighted:  # todo weighted not working, values multiplication does not work
+                base_counts_by_position[i, binary_base_representations] = counts  # * values
+            else:
+                base_counts_by_position[i, binary_base_representations] = counts
+            keys >>= 2
+
+        return base_counts_by_position
+
+    def effective_diversity(self, weighted=True):
+        """returns the effective base diversity of the barcode set by position.
+
+        maximum diversity for each position is 1, and represents a perfect split of 25% per base at
+        a given position.
+
+        :param bool weighted: (Default False) if True, each barcode is counted once for each time it
+          was observed
+
+        :return np.array[float]: array of size barcode_length containing floats in [0, 1]
+        """
+        return base4_entropy(self.base_frequency(weighted=weighted))
+
+
+class PriorBarcodeSet(BarcodeBase):
+    """Creates a set of barcodes from file
+
+    :method from_whitelist: construct a PriorBarcodeSet from a whitelist file
+    :method summarize_hamming_distances:  returns descriptive statistics on hamming distances
+      between pairs of barcodes
+    """
+
+    @classmethod
+    def from_whitelist(cls, file_, barcode_length):
+        """Creates a barcode set from a whitelist file
+
+        :param str file_: location of the whitelist file. Should be formatted one barcode per line.
```

Comment:
It looks like the barcodes are expected to be strings like 'ACGT', not two bit or three bit encoded, etc. Would be helpful to specify that or give an example barcode string, since elsewhere in this package we expect two bit or three bit encoded strings.

Reply:
Added comment in file_ docstring.